File: services/matching/normalise.py
"""Shared deterministic name normalization rules."""

from __future__ import annotations

import re
import unicodedata

LEGAL_SUFFIXES = sorted(
    [
        "f z l l c",
        "f z c o",
        "fz llc",
        "fzllc",
        "fzco",
        "fze",
        "dmcc",
        "establishment",
        "est",
        "international",
        "limited",
        "trading",
        "company",
        "gesellschaft mit beschrankter haftung",
        "gmbh",
        "holding",
        "group",
        "gen trdg",
        "intl",
        "pjsc",
        "psc",
        "wll",
        "spc",
        "s a",
        "b v",
        "sa",
        "bv",
        "llc",
        "l l c",
        "ltd",
        "corp",
        "inc",
        "co",
    ],
    key=len,
    reverse=True,
)

HONORIFICS = sorted(
    [
        "sheikh",
        "sh",
        "al haj",
        "haji",
        "dr",
        "eng",
        "mr",
        "mrs",
        "sayed",
        "sayyid",
        "prof",
        "h e",
    ],
    key=len,
    reverse=True,
)

PARTICLE_ALIASES = [
    ("bin", "ben", "ibn", "bn", "b"),
    ("abu", "abo", "abou", "aboo"),
    ("abd", "abdul", "abdel", "abdal", "abd al", "abd el"),
    ("umm", "om"),
]


def fold_diacritics(value: str) -> str:
    return "".join(
        character
        for character in unicodedata.normalize("NFKD", value)
        if unicodedata.category(character) != "Mn"
    )


def as_tokens(value: str) -> list[str]:
    return [token for token in value.split() if token]


def clean_text(value: str) -> str:
    output: list[str] = []
    for character in fold_diacritics(str(value or "")).lower():
        if character in "-'’`":
            output.append(" ")
        elif unicodedata.category(character)[0] in {"L", "N"}:
            output.append(character)
        else:
            output.append(" ")
    return " ".join(as_tokens("".join(output)))


def remove_whole_word_phrases(value: str, phrases: list[str]) -> str:
    output = value
    for phrase in phrases:
        pattern = r"\s+".join(re.escape(part) for part in phrase.split(" "))
        output = re.sub(rf"\b{pattern}\b", " ", output)
    return " ".join(as_tokens(output))


def normalise_name(raw_name: str) -> str:
    value = clean_text(raw_name)
    value = remove_whole_word_phrases(value, HONORIFICS)
    return remove_whole_word_phrases(value, LEGAL_SUFFIXES)


def replace_particle_aliases(tokens: list[str]) -> list[str]:
    output = tokens[:]
    for aliases in PARTICLE_ALIASES:
        alias_set = set(aliases)
        output = [aliases[0] if token in alias_set else token for token in output]
    return output


def generate_name_variants(raw_name: str) -> list[str]:
    base = normalise_name(raw_name)
    if not base:
        return []

    variants = {base}
    canonical_particles = replace_particle_aliases(as_tokens(base))
    variants.add(" ".join(canonical_particles))

    if canonical_particles and canonical_particles[0] in {"al", "el", "ul"}:
        variants.add(" ".join(canonical_particles[1:]))
        if len(canonical_particles) > 1:
            variants.add(f"{canonical_particles[0]}{''.join(canonical_particles[1:])}")

    first = canonical_particles[0] if canonical_particles else ""
    if first in {"abd", "abdul", "abdel", "abdal"} and len(canonical_particles) > 1:
        variants.add(" ".join(canonical_particles))
        variants.add("".join(canonical_particles[:2]) + (
            f" {' '.join(canonical_particles[2:])}" if len(canonical_particles) > 2 else ""
        ))

    for aliases in PARTICLE_ALIASES:
        if aliases[0] in canonical_particles:
            for alias in aliases:
                variants.add(
                    " ".join(alias if token == aliases[0] else token for token in canonical_particles)
                )
            variants.add(" ".join(token for token in canonical_particles if token != aliases[0]))

    return sorted(variant for variant in variants if variant)

File: jobs/shared/normalise.js
'use strict';

const LEGAL_SUFFIXES = [
  'f z l l c',
  'f z c o',
  'fz llc',
  'fzllc',
  'fzco',
  'fze',
  'dmcc',
  'establishment',
  'est',
  'international',
  'limited',
  'trading',
  'company',
  'gesellschaft mit beschrankter haftung',
  'gmbh',
  'holding',
  'group',
  'gen trdg',
  'intl',
  'pjsc',
  'psc',
  'wll',
  'spc',
  's a',
  'b v',
  'sa',
  'bv',
  'llc',
  'l l c',
  'ltd',
  'corp',
  'inc',
  'co'
].sort((a, b) => b.length - a.length);

const HONORIFICS = [
  'sheikh',
  'sh',
  'al haj',
  'haji',
  'dr',
  'eng',
  'mr',
  'mrs',
  'sayed',
  'sayyid',
  'prof',
  'h e'
].sort((a, b) => b.length - a.length);

const PARTICLE_ALIASES = [
  ['bin', 'ben', 'ibn', 'bn', 'b'],
  ['abu', 'abo', 'abou', 'aboo'],
  ['abd', 'abdul', 'abdel', 'abdal', 'abd al', 'abd el'],
  ['umm', 'om']
];

function foldDiacritics(value) {
  return value.normalize('NFKD').replace(/\p{M}/gu, '');
}

function asTokens(value) {
  return value.split(/\s+/).filter(Boolean);
}

function cleanText(value) {
  return asTokens(
    foldDiacritics(String(value || ''))
      .toLowerCase()
      .replace(/[-'’`]/g, ' ')
      .replace(/[^\p{L}\p{N}]+/gu, ' ')
  ).join(' ');
}

function removeWholeWordPhrases(value, phrases) {
  let output = value;
  for (const phrase of phrases) {
    const pattern = phrase
      .split(' ')
      .map((part) => part.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'))
      .join('\\s+');
    output = output.replace(new RegExp(`\\b${pattern}\\b`, 'g'), ' ');
  }
  return asTokens(output).join(' ');
}

function normaliseName(rawName) {
  let value = cleanText(rawName);
  value = removeWholeWordPhrases(value, HONORIFICS);
  value = removeWholeWordPhrases(value, LEGAL_SUFFIXES);
  return value;
}

function replaceParticleAliases(tokens) {
  return PARTICLE_ALIASES.reduce((result, aliases) => {
    const aliasSet = new Set(aliases);
    return result.map((token) => aliasSet.has(token) ? aliases[0] : token);
  }, tokens);
}

function generateNameVariants(rawName) {
  const base = normaliseName(rawName);
  if (!base) return [];

  const variants = new Set([base]);
  const tokens = asTokens(base);
  const canonicalParticles = replaceParticleAliases(tokens);
  variants.add(canonicalParticles.join(' '));

  if (/^(al|el|ul)$/.test(canonicalParticles[0])) {
    variants.add(canonicalParticles.slice(1).join(' '));
  }
  if (canonicalParticles.length > 1 && /^(al|el|ul)$/.test(canonicalParticles[0])) {
    variants.add(`${canonicalParticles[0]}${canonicalParticles.slice(1).join('')}`);
  }

  const first = canonicalParticles[0];
  if (first && /^(abd|abdul|abdel|abdal)$/.test(first) && canonicalParticles[1]) {
    variants.add([first, canonicalParticles[1], ...canonicalParticles.slice(2)].join(' '));
    variants.add([`${first}${canonicalParticles[1]}`, ...canonicalParticles.slice(2)].join(' '));
  }

  for (const aliases of PARTICLE_ALIASES) {
    if (canonicalParticles.some((token) => token === aliases[0])) {
      for (const alias of aliases) {
        variants.add(canonicalParticles.map((token) => token === aliases[0] ? alias : token).join(' '));
      }
      variants.add(canonicalParticles.filter((token) => token !== aliases[0]).join(' '));
    }
  }

  return [...variants].filter(Boolean);
}

function detectScript(value) {
  return /[\u0600-\u06ff]/u.test(String(value || '')) ? 'ARABIC' : 'LATIN';
}

function buildNameRecords(rawName, nameType = 'PRIMARY') {
  const variants = generateNameVariants(rawName);
  return variants.map((variant, index) => ({
    nameType: index === 0 ? nameType : 'AKA',
    script: detectScript(rawName),
    rawName: index === 0 ? String(rawName).trim() : variant,
    normalisedName: variant,
    nameTokens: asTokens(variant)
  }));
}

module.exports = {
  buildNameRecords,
  detectScript,
  generateNameVariants,
  normaliseName
};